pkg/core/klinedriver/bucket.go

```diff
@@ -0,0 +1,112 @@
+// Note that it will also update the bucket's status including the start/end time and the kline.
+func (bucket *Bucket) FillGapKLines(numShifts uint64) (kLines []*types.KLine) {
+	interval := bucket.Interval
+	if numShifts > 1 {
+		bucket.KLine.Open = bucket.KLine.Close
+		bucket.KLine.High = bucket.KLine.Close
+		bucket.KLine.Low = bucket.KLine.Close
+		bucket.KLine.Volume = fixedpoint.Zero
+		bucket.KLine.QuoteVolume = fixedpoint.Zero
+		bucket.KLine.TakerBuyBaseAssetVolume = fixedpoint.Zero
+		bucket.KLine.TakerBuyQuoteAssetVolume = fixedpoint.Zero
+		bucket.KLine.NumberOfTrades = 0
+	}
+	// fill the gap with empty trades
+	for i := uint64(1); i < numShifts; i++ {
+		bucket.StartTime = types.Time(bucket.StartTime.Time().Add(interval.Duration()))
+		bucket.EndTime = types.Time(bucket.EndTime.Time().Add(interval.Duration()))
+		bucket.KLine.StartTime = types.Time(bucket.StartTime.Time().UTC())
+		bucket.KLine.EndTime = types.Time(bucket.EndTime.Time().UTC())
+		// the opening kline
+		kline := *bucket.KLine // copy
+		kline.EndTime = types.Time(bucket.StartTime.Time().UTC().Add(time.Millisecond))
+		kline.Closed = false
+		kLines = append(kLines, &kline)
+		// the closing kline
+		kline = *bucket.KLine // copy again
+		kline.EndTime = types.Time(bucket.EndTime.Time().UTC())
+		kline.Closed = true
+		kLines = append(kLines, &kline)
+	}
+	return kLines
+}
+
+func (bucket *Bucket) accumulateTrade(trade *types.Trade) {
+	// expecting the trade time is in the same timezone as the kline
+	// if you are about to use the trade time, make sure to convert the timezone properly
+	// IMPORTANT: the trade should be treated as read-only here
+	tradeTimeUTC := types.Time(trade.Time.Time().UTC())
+	bucket.KLine.Exchange = trade.Exchange
+	bucket.KLine.EndTime = tradeTimeUTC
+	bucket.KLine.Close = trade.Price
+	bucket.KLine.High = fixedpoint.Max(bucket.KLine.High, trade.Price)
+	if bucket.KLine.NumberOfTrades == 0 {
+		bucket.KLine.Open = trade.Price
+		bucket.KLine.Low = trade.Price
+	} else {
+		bucket.KLine.Low = fixedpoint.Min(bucket.KLine.Low, trade.Price)
+	}
+	bucket.KLine.Volume = bucket.KLine.Volume.Add(trade.Quantity)
+	bucket.KLine.QuoteVolume = bucket.KLine.QuoteVolume.Add(trade.QuoteQuantity)
+	bucket.KLine.NumberOfTrades++
+	bucket.KLine.LastTradeID = trade.ID
+	if trade.IsBuyer && !trade.IsMaker {
+		bucket.KLine.TakerBuyBaseAssetVolume = bucket.KLine.TakerBuyBaseAssetVolume.Add(trade.Quantity)
+		bucket.KLine.TakerBuyQuoteAssetVolume = bucket.KLine.TakerBuyQuoteAssetVolume.Add(trade.QuoteQuantity)
+	}
+
+}
+
+func (bucket *Bucket) findNextBucket(currentTime types.Time) (nextBucket *Bucket, numShifts uint64) {
+	nextBucket = &Bucket{
+		StartTime: bucket.StartTime,
+		EndTime:   bucket.EndTime,
+		Interval:  bucket.Interval,
+		KLine:     nil,
+	}
+	if currentTime.Before(nextBucket.StartTime.Time()) {
+		return nextBucket, numShifts
+	}
+	for {
```

Comment:
seems like we could directly calculate the nextBucket StartTime and EndTime from Interval and currentTime.
Are you able to rewrite this to a non-loop version?

Reply:
Good point.
Working on it.

Reply:
Done.
I also add a unit test on it.
Please have a look.